controllers/barbican_controller.go

```diff
@@ -1152,3 +1153,26 @@ func (r *BarbicanReconciler) ensureDB(
 	instance.Status.Conditions.MarkTrue(condition.DBReadyCondition, condition.DBReadyMessage)
 	return db, ctrlResult, nil
 }
+
+func cleanupOldDeployment(ctx context.Context, c client.Client, namespace string, oldName string) error {
+	var oldDeployment appsv1.Deployment
+	err := c.Get(ctx, types.NamespacedName{Name: oldName, Namespace: namespace}, &oldDeployment)
+	if err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	if oldDeployment.Spec.Replicas == nil || *oldDeployment.Spec.Replicas > 0 {
+		replicas := int32(0)
+		oldDeployment.Spec.Replicas = &replicas
+		if updateErr := c.Update(ctx, &oldDeployment); updateErr != nil {
+			return fmt.Errorf("failed to scale down old deployment '%s': %w", oldName, updateErr)
+		}
+		return nil
+	}
+
+	if delErr := c.Delete(ctx, &oldDeployment); delErr != nil {
+		return fmt.Errorf("failed to delete old deployment '%s': %w", oldName, delErr)
+	}
+
```

Comment:
wouldn't it be enough to only run the delete and ignore the `NotFound`, which would be a single api call, instead of 3 ?
```
	deployment := &appsv1.Deployment{
		ObjectMeta: metav1.ObjectMeta{
			Name:      oldName,
			Namespace: namespace,
		},
	}

	err := c.Delete(ctx, deployment)
	if err != nil && !k8s_errors.IsNotFound(err) {
		return fmt.Errorf("Error cleanup old deployment %s: %w", oldName, err)
	}
```

Reply:
in a later release, when we are comfortable that users were on a release with the cleanup, could remove it again.